### ISSUE TITLE
Controller reconciling indefinitely after applying sbr yaml

### ISSUE THREAD

**Issue by Avni-Sharma** [redhat-developer/service-binding-operator#509]
After #442 got merged the controller has started to watch on application workloads. When we add a SBR to connect the app and db the reconcile is on an indefinite loop, whereas it should be stimulus to changes in the sbr yaml. Now if we see the topology view(OpenShift 4.x +) the app pod dougnout indefinitely changes itself from lighter blue to darker blue circles. The activity has been captured here 
https://bluejeans.com/s/jli5w/
I am thinking it is because the secret is getting injected again and again, and because we are watching the workload the reconciler hits and it is like a chicken egg prob. The first time a secret is injected, the sbr controller reads it as a change and reconciles again, and again secret is injected and this goes on.

**Comment by Avni-Sharma** [comment]
Another issue that I am facing is that the resourceVersion in the secret is changing even when data, annotations is not changing.
Old secret
```
avni@localhost service-binding-operator (reconcile-indefinte)=>>   k get secrets binding-request -o yaml
apiVersion: v1
data:
  DATABASE_CONFIGMAP_DB_HOST: MTcyLjMwLjYxLjI0Mw==
  DATABASE_CONFIGMAP_DB_NAME: ZGItZGVtbw==
  DATABASE_CONFIGMAP_DB_PASSWORD: cGFzc3dvcmQ=
  DATABASE_CONFIGMAP_DB_PORT: NTQzMg==
  DATABASE_DBCONNECTIONIP: MTcyLjMwLjYxLjI0Mw==
  DATABASE_DBCONNECTIONPORT: NTQzMg==
  DATABASE_DBNAME: ZGItZGVtbw==
  DATABASE_SECRET_PASSWORD: cGFzc3dvcmQ=
  DATABASE_SECRET_USER: cG9zdGdyZXM=
kind: Secret
metadata:
  annotations:
    service-binding-operator.apps.openshift.io/binding-name: binding-request
    service-binding-operator.apps.openshift.io/binding-namespace: default
  creationTimestamp: "2020-06-22T05:50:45Z"
  managedFields:
  - apiVersion: v1
    fieldsType: FieldsV1
    fieldsV1:
      f:data:
        .: {}
        f:DATABASE_CONFIGMAP_DB_HOST: {}
        f:DATABASE_CONFIGMAP_DB_NAME: {}
        f:DATABASE_CONFIGMAP_DB_PASSWORD: {}
        f:DATABASE_CONFIGMAP_DB_PORT: {}
        f:DATABASE_DBCONNECTIONIP: {}
        f:DATABASE_DBCONNECTIONPORT: {}
        f:DATABASE_DBNAME: {}
        f:DATABASE_SECRET_PASSWORD: {}
        f:DATABASE_SECRET_USER: {}
      f:metadata:
        f:annotations:
          .: {}
          f:service-binding-operator.apps.openshift.io/binding-name: {}
          f:service-binding-operator.apps.openshift.io/binding-namespace: {}
      f:type: {}
    manager: service-binding-operator-local
    operation: Update
    time: "2020-06-22T08:53:35Z"
  name: binding-request
  namespace: default
  resourceVersion: "161806"
  selfLink: /api/v1/namespaces/default/secrets/binding-request
  uid: 28246bf6-1733-4056-b51c-a3ac74d95f2d
type: Opaque
```

New secret
```
avni@localhost service-binding-operator (reconcile-indefinte)=>>   k get secrets binding-request -o yaml
apiVersion: v1
data:
  DATABASE_CONFIGMAP_DB_HOST: MTcyLjMwLjYxLjI0Mw==
  DATABASE_CONFIGMAP_DB_NAME: ZGItZGVtbw==
  DATABASE_CONFIGMAP_DB_PASSWORD: cGFzc3dvcmQ=
  DATABASE_CONFIGMAP_DB_PORT: NTQzMg==
  DATABASE_DBCONNECTIONIP: MTcyLjMwLjYxLjI0Mw==
  DATABASE_DBCONNECTIONPORT: NTQzMg==
  DATABASE_DBNAME: ZGItZGVtbw==
  DATABASE_SECRET_PASSWORD: cGFzc3dvcmQ=
  DATABASE_SECRET_USER: cG9zdGdyZXM=
kind: Secret
metadata:
  annotations:
    service-binding-operator.apps.openshift.io/binding-name: binding-request
    service-binding-operator.apps.openshift.io/binding-namespace: default
  creationTimestamp: "2020-06-22T05:50:45Z"
  managedFields:
  - apiVersion: v1
    fieldsType: FieldsV1
    fieldsV1:
      f:data:
        .: {}
        f:DATABASE_CONFIGMAP_DB_HOST: {}
        f:DATABASE_CONFIGMAP_DB_NAME: {}
        f:DATABASE_CONFIGMAP_DB_PASSWORD: {}
        f:DATABASE_CONFIGMAP_DB_PORT: {}
        f:DATABASE_DBCONNECTIONIP: {}
        f:DATABASE_DBCONNECTIONPORT: {}
        f:DATABASE_DBNAME: {}
        f:DATABASE_SECRET_PASSWORD: {}
        f:DATABASE_SECRET_USER: {}
      f:metadata:
        f:annotations:
          .: {}
          f:service-binding-operator.apps.openshift.io/binding-name: {}
          f:service-binding-operator.apps.openshift.io/binding-namespace: {}
      f:type: {}
    manager: service-binding-operator-local
    operation: Update
    time: "2020-06-22T09:02:49Z"
  name: binding-request
  namespace: default
  resourceVersion: "170950"
  selfLink: /api/v1/namespaces/default/secrets/binding-request
  uid: 28246bf6-1733-4056-b51c-a3ac74d95f2d
type: Opaque
```

**Comment by pedjak** [comment]
@Avni-Sharma can you post here the complete list of steps for reproducing this?

**Comment by Avni-Sharma** [comment]
Steps to reproduce- 
- Create an imported app
- Create a service
- Apply sbr yaml
(the above 3 steps can be referred from https://github.com/redhat-developer/service-binding-operator/tree/master/examples/nodejs_postgresql)

Now when i do `kubectl get pods` I see an indefinite creation and termination of nodejs app pods

```
avni@localhost service-binding-operator (reconcile-indefinte)=>>   k get deployments
NAME                    READY   UP-TO-DATE   AVAILABLE   AGE
db-demo-postgresql      1/1     1            1           4h14m
nodejs-rest-http-crud   1/1     1            1           4h15m
avni@localhost service-binding-operator (reconcile-indefinte)=>>   k get pods
NAME                                     READY   STATUS        RESTARTS   AGE
db-demo-postgresql-7dccb8b655-hlm94      1/1     Running       0          4h14m
nodejs-rest-http-crud-1-build            0/1     Completed     0          4h15m
nodejs-rest-http-crud-5695d4f9db-w2f5b   1/1     Running       0          4s
nodejs-rest-http-crud-7ddb59c98f-jz4gc   1/1     Terminating   0          9s
avni@localhost service-binding-operator (reconcile-indefinte)=>>   k get pods
NAME                                     READY   STATUS        RESTARTS   AGE
db-demo-postgresql-7dccb8b655-hlm94      1/1     Running       0          4h14m
nodejs-rest-http-crud-1-build            0/1     Completed     0          4h15m
nodejs-rest-http-crud-5695d4f9db-w2f5b   1/1     Terminating   0          9s
nodejs-rest-http-crud-7dbfb6875c-szzsk   1/1     Running       0          4s


nodejs-rest-http-crud-7dbfb6875c-szzsk   1/1     Running       0          4s
avni@localhost service-binding-operator (reconcile-indefinte)=>>   k get pods
NAME                                     READY   STATUS              RESTARTS   AGE
db-demo-postgresql-7dccb8b655-hlm94      1/1     Running             0          4h16m
image-hooks-http-6cb49b968c-t22dl        0/1     ContainerCreating   0          54s
nodejs-rest-http-crud-1-build            0/1     Completed           0          4h17m
nodejs-rest-http-crud-686ffd8958-bt2fg   0/1     Terminating         0          10s
nodejs-rest-http-crud-7777464d8b-9dwmm   1/1     Running             0          5s
avni@localhost service-binding-operator (reconcile-indefinte)=>>   k get pods
NAME                                     READY   STATUS              RESTARTS   AGE
db-demo-postgresql-7dccb8b655-hlm94      1/1     Running             0          4h16m
image-hooks-http-6cb49b968c-t22dl        0/1     ContainerCreating   0          56s
nodejs-rest-http-crud-1-build            0/1     Completed           0          4h17m
nodejs-rest-http-crud-5b665dfd75-6n78g   0/1     ContainerCreating   0          2s
nodejs-rest-http-crud-686ffd8958-bt2fg   0/1     Terminating         0          12s
nodejs-rest-http-crud-7777464d8b-9dwmm   1/1     Running             0          7s
```

**Comment by isutton** [comment]
It seems the `time` property in the slice of managed fields is updated, and this information should be taken in consideration when making a decision whether a resource should or not trigger a service binding reconciliation.

> ```
> metadata:
>   managedFields:
>   - apiVersion: v1
>     manager: service-binding-operator-local
>     operation: Update
>     time: "2020-06-22T08:53:35Z"
> ```

Because of the composition of an entry (manager vs operation vs time) it is required to understand how those fields are updated by the API server and how to properly react when this happens.

Going a little bit more in depth, the `metadata.managedFields` is related to [Server Side Apply](https://kubernetes.io/docs/reference/using-api/api-concepts/#server-side-apply), which has been enabled in Kubernetes since 1.16 [according to this document](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/) which makes this bug quite important.

In my setup, running CRC 1.11, it seems the feature is disabled as I couldn't find any evidence in metadata of existing Secrets or ConfigMaps in the cluster, even if the Kubernetes API being served being v1.17.1 as seen below:

```
oc version       
Client Version: v4.2.0-alpha.0-526-g38dae9c
Server Version: 4.4.5
Kubernetes Version: v1.17.1
```

@Avni-Sharma, what does `oc version` say in your setup?

**Comment by Avni-Sharma** [comment]
Hi @isutton . Thanks for this, managedFields is something new that I noticed
```
avni@localhost service-binding-operator (reconcile-indefinte)=>>   oc version
oc v3.11.0+0cbc58b
kubernetes v1.11.0+d4cacc0
features: Basic-Auth GSSAPI Kerberos SPNEGO

Server https://api.dev-svc-4.5-062204.devcluster.openshift.com:6443
kubernetes v1.18.3+a637491
```

**Comment by Avni-Sharma** [comment]
Also maybe the feature is disabled in crc. [many things are disabled in crc by default]
https://kubernetes.io/docs/reference/using-api/api-concepts/#disabling-the-feature
`--feature-gates ServerSideApply=false`

**Comment by sbose78** [comment]
Note,
Openshift 4.5 has `managedFields` 'enabled'.
OpenShift 4.4 does not have `managedFields`.


Is this issue visible in a 4.4 cluster?

**Comment by Avni-Sharma** [comment]
@isutton  Should we consider overriding field managers with an empty string? while updating the interim secret https://github.com/redhat-developer/service-binding-operator/blob/master/pkg/controller/servicebindingrequest/secret.go#L55
I see only the FieldManager in the UpdateOptions but not `time`. :thinking:

**Comment by sbose78** [comment]
I had seen the infinite reconcile issue almost 8-9 months back when I had disabled it. The controller was trying to update the annotations in the Deployment object and was failing to do so because the local copy of the Deployment object was going stale.

**Comment by Avni-Sharma** [comment]
Hi @sbose78 
Tried on 4.4 cluster and resourceVersion is changing for 4.4 cluster as well, regardless of the managedFields

old secret
```
avni@localhost service-binding-operator (reconcile-indefinte)=>>   k get secrets binding-request -o yaml
apiVersion: v1
data:
  DATABASE_CONFIGMAP_DB_HOST: MTcyLjMwLjc2LjMy
  DATABASE_CONFIGMAP_DB_NAME: ZGItZGVtbw==
  DATABASE_CONFIGMAP_DB_PASSWORD: cGFzc3dvcmQ=
  DATABASE_CONFIGMAP_DB_PORT: NTQzMg==
  DATABASE_DBCONNECTIONIP: MTcyLjMwLjc2LjMy
  DATABASE_DBCONNECTIONPORT: NTQzMg==
  DATABASE_DBNAME: ZGItZGVtbw==
  DATABASE_SECRET_PASSWORD: cGFzc3dvcmQ=
  DATABASE_SECRET_USER: cG9zdGdyZXM=
kind: Secret
metadata:
  annotations:
    service-binding-operator.apps.openshift.io/binding-name: binding-request
    service-binding-operator.apps.openshift.io/binding-namespace: default
  creationTimestamp: "2020-06-22T20:41:09Z"
  name: binding-request
  namespace: default
  resourceVersion: "224202"
  selfLink: /api/v1/namespaces/default/secrets/binding-request
  uid: 7e03edcd-10db-4a95-945e-d52ef4b0e178
type: Opaque
```

new secret 
```
avni@localhost service-binding-operator (reconcile-indefinte)=>>   k get secrets binding-request -o yaml
apiVersion: v1
data:
  DATABASE_CONFIGMAP_DB_HOST: MTcyLjMwLjc2LjMy
  DATABASE_CONFIGMAP_DB_NAME: ZGItZGVtbw==
  DATABASE_CONFIGMAP_DB_PASSWORD: cGFzc3dvcmQ=
  DATABASE_CONFIGMAP_DB_PORT: NTQzMg==
  DATABASE_DBCONNECTIONIP: MTcyLjMwLjc2LjMy
  DATABASE_DBCONNECTIONPORT: NTQzMg==
  DATABASE_DBNAME: ZGItZGVtbw==
  DATABASE_SECRET_PASSWORD: cGFzc3dvcmQ=
  DATABASE_SECRET_USER: cG9zdGdyZXM=
kind: Secret
metadata:
  annotations:
    service-binding-operator.apps.openshift.io/binding-name: binding-request
    service-binding-operator.apps.openshift.io/binding-namespace: default
  creationTimestamp: "2020-06-22T20:41:09Z"
  name: binding-request
  namespace: default
  resourceVersion: "224326"
  selfLink: /api/v1/namespaces/default/secrets/binding-request
  uid: 7e03edcd-10db-4a95-945e-d52ef4b0e178
type: Opaque
```

**Comment by isutton** [comment]
> Also maybe the feature is disabled in crc. [many things are disabled in crc by default]
> https://kubernetes.io/docs/reference/using-api/api-concepts/#disabling-the-feature
> `--feature-gates ServerSideApply=false`

@Avni-Sharma the feature is indeed disabled in CRC 1.11 because it distributes OpenShift 4.4, thanks for confirming it.

> @isutton Should we consider overriding field managers with an empty string? while updating the interim secret https://github.com/redhat-developer/service-binding-operator/blob/master/pkg/controller/servicebindingrequest/secret.go#L55
> I see only the FieldManager in the UpdateOptions but not `time`. 🤔

@Avni-Sharma: my current understanding is that `metadata.managedFields` is used to hold data that is changed by different actors to allow server-side logic to properly mutate the resource, so in this particular case the resource is indeed being modified to keep track of server-side apply which can be demonstrated by the `time` property that has been changed in between runs.

In fact, this tells us the operator is indiscriminately updating `data` and `metadata.annotations` at each reconciliation; this can definitely be avoided by checking, before updating the secret in `secret.createOrUpdate()`, whether the operation *should* indeed be executed.

The verification is performed by changing the current strategy from _attempt to create the secret and in case it exists, update it with the new secret contents_ to instead _attempt to get the expected secret and update it in case it exists, otherwise create it_ (from L49 until L58):

https://github.com/redhat-developer/service-binding-operator/blob/ea4ecd9dac0e4c16ab37ebf871ca1ffe8d63e8e9/pkg/controller/servicebindingrequest/secret.go#L30-L60

**Comment by Avni-Sharma** [comment]
Even if there is an Update() call and the data payload has not changed, I don't think the secret Object should get updated. It is just like me trying to applying a secret.yaml file which has already been applied and then I get `secret/mysecret unchanged`  in this case the resorceVersion should not change even if the update() has been called

**Comment by Avni-Sharma** [comment]
On observation, I see that the env[0].value keeps changing in the spec and resulting in an overall reconcile again even when the secret's resourceVersion issue has been handled in the WIP PR - 523
```
    spec:
      containers:
      - env:
        - name: ServiceBindingOperatorChangeTriggerEnvVar
          value: "2020-06-28T21:31:10+05:30"
        envFrom:
        - secretRef:
            name: binding-request

```


```
     containers:
      - env:
        - name: ServiceBindingOperatorChangeTriggerEnvVar
          value: "2020-06-28T21:33:10+05:30"
        envFrom:
        - secretRef:
            name: binding-request
```

**Comment by Avni-Sharma** [comment]
Fix got merged, closing the issue
